cmd/soroban-rpc/internal/db/migration.go

```diff
@@ -165,55 +181,69 @@ func (g *guardedMigration) Commit(ctx context.Context) error {
 	return setMetaBool(ctx, g.db, g.guardMetaKey, true)
 }
 
-func GetMigrationLedgerRange(ctx context.Context, db *DB, retentionWindow uint32) (*LedgerSeqRange, error) {
+func GetMigrationLedgerRange(ctx context.Context, db *DB, retentionWindow uint32) (LedgerSeqRange, error) {
 	firstLedgerToMigrate := firstLedger
 	latestLedger, err := NewLedgerEntryReader(db).GetLatestLedgerSequence(ctx)
 	if err != nil && !errors.Is(err, ErrEmptyDB) {
-		return nil, fmt.Errorf("failed to get latest ledger sequence: %w", err)
+		return LedgerSeqRange{}, fmt.Errorf("failed to get latest ledger sequence: %w", err)
 	}
 	if latestLedger > retentionWindow {
 		firstLedgerToMigrate = latestLedger - retentionWindow
 	}
-	return &LedgerSeqRange{
+	return LedgerSeqRange{
 		First: firstLedgerToMigrate,
 		Last:  latestLedger,
 	}, nil
 }
 
-func BuildMigrations(ctx context.Context, logger *log.Entry, db *DB, networkPassphrase string,
-	ledgerSeqRange *LedgerSeqRange,
-) (MultiMigration, error) {
+func BuildMigrations(
+	ctx context.Context, logger *log.Entry, db *DB, networkPassphrase string,
+	ledgerSeqRange LedgerSeqRange,
+) (MultiMigration, LedgerSeqRange, error) {
+	// Track ranges for which migrations are actually necessary
+	applicableRange := LedgerSeqRange{}
+
 	// Start a common db transaction for the entire migration duration
 	err := db.Begin(ctx)
 	if err != nil {
-		return MultiMigration{}, errors.Join(err, db.Rollback())
+		return MultiMigration{}, applicableRange, errors.Join(err, db.Rollback())
 	}
 
-	migrationNameToFunc := map[string]migrationApplierF{
+	//
+	// Add new migrations here:
+	//
+	currentMigrations := map[string]migrationApplierF{
 		transactionsMigrationName: newTransactionTableMigration,
 		eventsMigrationName:       newEventTableMigration,
 	}
 
-	migrations := make([]Migration, 0, len(migrationNameToFunc))
-
-	for migrationName, migrationFunc := range migrationNameToFunc {
+	migrations := make([]Migration, 0, len(currentMigrations))
+	for migrationName, migrationFunc := range currentMigrations {
 		migrationLogger := logger.WithField("migration", migrationName)
 		factory := migrationFunc(
 			ctx,
 			migrationLogger,
 			networkPassphrase,
-			ledgerSeqRange,
+			&ledgerSeqRange,
 		)
 
 		guardedM, err := newGuardedDataMigration(ctx, migrationName, migrationLogger, factory, db)
 		if err != nil {
-			return MultiMigration{}, errors.Join(err, fmt.Errorf(
+			return MultiMigration{}, applicableRange, errors.Join(err, fmt.Errorf(
 				"could not create guarded migration for %s", migrationName), db.Rollback())
 		}
+
+		if guardedM == nil {
```

Comment:
Since we added the check here, I think we can remove all the `if g.alreadyMigrated` conditions in `guardedMigration` interface - https://github.com/stellar/soroban-rpc/blob/main/cmd/soroban-rpc/internal/db/migration.go#L142

Reply:
Technically yes, since we won't run that code path the way it's written now, but I'd rather leave it in to keep each abstraction layer fully functional within itself.